### ISSUE TITLE
Improper Values in getTranslation

### ISSUE THREAD

**Issue by deep4U** [gameplay3d/GamePlay#1834]
Hi
So I load and create a  scene and what I do in Initialize 

Scene* sce= bundle->loadScene();
nodeengein = sce->getFirstNode();//->loadNode("trainengroot");
nodeengein->scale(0.2f);
nodeengein->translate(-1.0f,0.3f,0.7f);

Now In render function In an attempt to know effect on speed of objects on frame drop due to gpu load I tried this

gameplay::print("nodeengein translation is %f ============ \n",nodeengein->getTranslation().x);
struct timeval now;
gettimeofday(&now, NULL);
if(nodeengein->getTranslation().x==-1.0f )
gameplay::print(" %ld sec %ld usec rech 0 at",now.tv_sec,now.tv_usec);
else if(nodeengein->getTranslation().x==-5.0f)
gameplay::print(" %ld sec %ld usec going to 0 from 1 at",now.tv_sec,now.tv_usec);
nodeengein->translate(-0.01f,0.0f,0.0f);

And I get values like 

nodeengein translation is -1.010000 ============ 
nodeengein translation is -1.020000 ============ 
nodeengein translation is -1.030000 ============ 
nodeengein translation is -1.040000 ============ 
so on keeps reducing my 0.01 as expected
nodeengein translation is -1.500000 ============ 
nodeengein translation is -1.510000 ============ 
nodeengein translation is -1.520000 ============ 
and suddenly !!!!!
nodeengein translation is -1.529999 ============ 
nodeengein translation is -1.539999 ============ 
nodeengein translation is -1.549999 ============ 
nodeengein translation is -1.559999 ============ 
nodeengein translation is -1.569999 ============ 
nodeengein translation is -1.579999 ============ 
nodeengein translation is -1.589999 ============ 
nodeengein translation is -1.599999 ============ 
nodeengein translation is -1.609999 ============ 
nodeengein translation is -1.619999 ============ 
nodeengein translation is -1.629999 ============ 
nodeengein translation is -1.639999 ============ 
nodeengein translation is -1.649999 ============ 
nodeengein translation is -1.659999 ============ 
nodeengein translation is -1.669999 ============ 
nodeengein translation is -1.679999 ============ 
nodeengein translation is -1.689999 ============ 
nodeengein translation is -1.699999 ============ 
nodeengein translation is -1.709999 ============ 
nodeengein translation is -1.719999 ============ 
nodeengein translation is -1.729999 ============ 
nodeengein translation is -1.739999 ============ 
nodeengein translation is -1.749999 ============ 
nodeengein translation is -1.759999 ============ 
nodeengein translation is -1.769999 ============ 
nodeengein translation is -1.779999 ============ 
nodeengein translation is -1.789999 ============ 
nodeengein translation is -1.799999 ============ 
nodeengein translation is -1.809999 ============ 
nodeengein translation is -1.819999 ============ 
nodeengein translation is -1.829999 ============ 
nodeengein translation is -1.839999 ============ 
nodeengein translation is -1.849999 ============ 
nodeengein translation is -1.859999 ============ 
nodeengein translation is -1.869999 ============ 
nodeengein translation is -1.879999 ============ 
nodeengein translation is -1.889999 ============ 
nodeengein translation is -1.899999 ============ 
nodeengein translation is -1.909999 ============ 
nodeengein translation is -1.919999 ============ 
nodeengein translation is -1.929999 ============ 
nodeengein translation is -1.939999 ============ 
nodeengein translation is -1.949999 ============ 
nodeengein translation is -1.959999 ============ 
nodeengein translation is -1.969999 ============ 
nodeengein translation is -1.979999 ============ 
nodeengein translation is -1.989999 ============ 
nodeengein translation is -1.999999 ============ 
nodeengein translation is -2.009999 ============ 
nodeengein translation is -2.019999 ============ 
nodeengein translation is -2.029999 ============ 
nodeengein translation is -2.039999 ============ 
nodeengein translation is -2.049999 ============ 
nodeengein translation is -2.059999 ============ 
nodeengein translation is -2.069999 ============ 
nodeengein translation is -2.079999 ============ 
nodeengein translation is -2.089999 ============ 
nodeengein translation is -2.099999 ============ 
nodeengein translation is -2.109999 ============ 
nodeengein translation is -2.119999 ============ 
nodeengein translation is -2.129999 ============ 
nodeengein translation is -2.139999 ============ 
nodeengein translation is -2.149999 ============ 
nodeengein translation is -2.159999 ============ 
nodeengein translation is -2.169999 ============ 
nodeengein translation is -2.179999 ============ 
nodeengein translation is -2.189999 ============ 
nodeengein translation is -2.199999 ============ 
nodeengein translation is -2.209999 ============ 
nodeengein translation is -2.219999 ============ 
nodeengein translation is -2.229999 ============ 
nodeengein translation is -2.239999 ============ 
nodeengein translation is -2.249999 ============ 
nodeengein translation is -2.259999 ============ 
nodeengein translation is -2.269999 ============ 
nodeengein translation is -2.279999 ============ 
nodeengein translation is -2.289999 ============ 
nodeengein translation is -2.299999 ============ 
nodeengein translation is -2.309999 ============ 
nodeengein translation is -2.319999 ============ 
nodeengein translation is -2.329999 ============ 
nodeengein translation is -2.339999 ============ 
nodeengein translation is -2.349999 ============ 
nodeengein translation is -2.359999 ============ 
nodeengein translation is -2.369999 ============ 
nodeengein translation is -2.379999 ============ 
nodeengein translation is -2.389999 ============ 
nodeengein translation is -2.399999 ============ 
nodeengein translation is -2.409999 ============ 
nodeengein translation is -2.419999 ============ 
nodeengein translation is -2.429999 ============ 
nodeengein translation is -2.439999 ============ 
nodeengein translation is -2.449999 ============ 
nodeengein translation is -2.459999 ============ 
nodeengein translation is -2.469999 ============ 
nodeengein translation is -2.479999 ============ 
nodeengein translation is -2.489999 ============ 
nodeengein translation is -2.499999 ============ 
nodeengein translation is -2.509999 ============ 
nodeengein translation is -2.519999 ============ 
nodeengein translation is -2.529999 ============ 
nodeengein translation is -2.539999 ============ 
nodeengein translation is -2.549999 ============ 
nodeengein translation is -2.559999 ============ 
nodeengein translation is -2.569999 ============ 
nodeengein translation is -2.579998 ============ 
nodeengein translation is -2.589998 ============ 
nodeengein translation is -2.599998 ============ 
nodeengein translation is -2.609998 ============ 
nodeengein translation is -2.619998 ============ 
nodeengein translation is -2.629998 ============ 
nodeengein translation is -2.639998 ============ 
nodeengein translation is -2.649998 ============ 
nodeengein translation is -2.659998 ============ 
nodeengein translation is -2.669998 ============ 
nodeengein translation is -2.679998 ============ 

even got 
-5.910041 afterwards
Is that a bug as I translate by 0.01 only and I got values from getTranslation that show it is not exact or is there some precision up-to which only it is perfect ?

**Comment by deep4U** [comment]
I mean only 0.01 is expected as difference between two consecutive values than how its coming like this

**Comment by xthydra** [comment]
that's funny cuz when i wrote a function that randomly generate objects position base on their bounding box, i spent like 4 hours trying to figure out why did the objects positions didn't seem to be consistent based on their world position Y but then i gave up on trying to figure out the bug.

**Comment by marcinjakubowski** [comment]
Floating point numbers are not precise and you should not use == to compare them. Have a look [here](http://floating-point-gui.de/errors/comparison/) for more information on comparing and floats in general.

**Comment by deep4U** [comment]
@marcinjakubowski Thanks
Thats ok 
but what I am talking about is the print above the if conditions 
The output of getTranslation is funky as from -1.0f no matter how many times I add 0.01 
Last digits should be always 0 only.
What I am worried about is weather the bug is in getTranslation or translate function as that would mean incorrect calculations for apps that depend on distance too much for example racing game.
@xthydra Can you confirm If that is the reason you gave up trying as it was only getTranslation based or are we having issue in translation of objects itself?

**Comment by marcinjakubowski** [comment]
@deep4U No, the last digits should not always be 0, due to how floats are stored. [Here's a short demo program] I just wrote to demonstrate it, pure C++

**Comment by deep4U** [comment]
@marcinjakubowski  you are right .....
Basically float uses only 24 bits for values after the decimal points so it is not precise(precision is upto 7 /8 digits I think)
Double however is more precise and is having 15 digit precision but uses 64 bits 
So now I am thinking how exactly do racing games  translate speed /distances to opengl coordinates accurately ( is it they use double or is it that they ignore this small error and few people lose the race due to float reducing their speed sometimes and blame themselves instead)

**Comment by lishuren** [comment]
Just my two cents, if I am wrong please correct me.
1 Float, double or decimal. They are totally different. You might be aware if your data structure could support your algorithm.
2 Racing games, or any other games, you will definitely need your game logic model. Node class is an engine class only. It is impossible to support all game logic requirements.
3 As you might be a game maker, you might need be aware of hardware capability. GPU is using float basically for better performance. Node's Translation will be passed to Shader. Think about how many registers you need. I believe float is the best practice so for.
